web/app.js

```diff
@@ -1542,6 +1542,33 @@ const PDFViewerApplication = {
     };
     internalEvents.set("updatefromsandbox", updateFromSandbox);
 
+    const visitedPages = new Set();
+    const pageOpen = async ({ pageNumber }) => {
+      let actions = null;
+      if (!visitedPages.has(pageNumber)) {
+        // Get actions only the first time the page is visited
+        const page = await pdfDocument.getPage(pageNumber);
+        actions = await page.getJSActions();
+        visitedPages.add(pageNumber);
+      }
```

Comment:
If you want to prevent *some* ordering issues, since this event-handler is currently asynchronous in some cases and not in others, something like the following could work I think:
```suggestion
    const visitedPages = new Map();
    const pageOpen = async ({ pageNumber }) => {
      if (!visitedPages.has(pageNumber)) {
        // Get actions only the first time the page is visited.
        const promise = new Promise(resolve => {
          const pageView = this.pdfViewer.getPageView(/* index = */ pageNumber - 1);
          if (pageView?.pdfPage) {
            pageView.pdfPage.getJSActions().then(resolve);
            return;
          }
          pdfDocument.getPage(pageNumber).then(pdfPage => {
            pdfPage.getJSActions().then(resolve);
          });
        });
        visitedPages.set(pageNumber, promise);
      }
      const actions = await visitedPages.get(pageNumber);
      
      if (pdfDocument !== this.pdfDocument) {
        return; // The document was closed while the actions resolved.
      }
```

---
This, however, still leaves the issue of all of this possibly happening *after* the associated "pageclose" event though.

Reply:
Thx for the patch: I almost copied/pasted it.
My goal was to not send actions when a page was visited again, so I just adapted your patch to this.

Reply:
> My goal was to not send actions when a page was visited again,

But why does sending it again matter though? You've already added code that should ignore that case in https://github.com/mozilla/pdf.js/pull/12816/files#diff-cd84875e4422389162fdb6971a6766a7666b89af6389e9039560ac32b342fae5R120-R122

The "issue" with the latest patch is, as far as I'm concerned, is that it's a lot less readable/understandable overall!

Reply:
Agreed, but at least for the quickjs version, we need to serialize/deserialize all the data which are evaluated in the sandbox.
So if we've some big scripts (do we irl ??), it's time/resource consuming.

Reply:
> Agreed, but at least for the quickjs version, we need to serialize/deserialize all the data which are evaluated in the sandbox.

I'm not overly worried about *that* part, as long as the mozilla-central version works OK; and I do think that we need easily readable code here (since otherwise it's too difficult to reason about it).

However, can you then instead please do what I suggested in my comments above/below, but with e.g. the following addendum here (it's possible this can become even shorter and more compact, but this should do the trick):
```javascript
const actions = await visitedPages.get(pageNumber);
if (actions) {
  // Avoid sending, and thus serializing, the `actions` data for each event.
  visitedPages.set(pageNumber, Promise.resolve(null));
}
```